Standard deviation tutorial

JakeSCahill opened the pull request:
- Converted the code to work with Serverless clusters.
- Added a tutorial, using the content of the README.md file.

JakeSCahill commented on compute/standard-deviation/client/src/main/java/your/company/name/ApplicationConfig.java:
```diff
@@ -0,0 +1,41 @@
+        clientConfig.getNetworkConfig().getCloudConfig()
+            .setDiscoveryToken(clusterDiscoveryToken)
+            .setEnabled(true);
+        clientConfig.setProperty("hazelcast.client.cloud.url", "https://uat.hazelcast.cloud");
```
```suggestion
```

angelasimms commented on compute/standard-deviation/client/src/main/java/your/company/name/Application.java:
```diff
@@ -3,20 +3,22 @@
+ * <p>The entry point of the application, which does the following
  * </p>
+ * <ol>
+ *    <li> Ask the user of the commond prompt for the
```
```suggestion
 *    <li> Asks the user of the command prompt for the
```

JakeSCahill replied:
```suggestion
 *    <li> Ask the user of the command prompt for the
```

angelasimms commented on compute/standard-deviation/client/src/main/java/your/company/name/SeniorDeveloper.java:
```diff
@@ -38,21 +40,24 @@ public class SeniorDeveloper {
-     * runs on each Hazelcast server process and returns a
-     * sub-total based on that server's data.
+     * runs on each Hazelcast member and returns a
+     * sub-total based on that members's data.
```
```suggestion
     * subtotal based on that members's data.
```

angelasimms commented on compute/standard-deviation/client/src/main/java/your/company/name/SeniorDeveloper.java:
```diff
@@ -62,8 +67,7 @@ public static double totalDifferenceSquared(IMap<Integer, Customer> iMap, double
 
-        /* Run the callable on all servers in parallel.
-         */
+        // Run the callable on all members in parallel.
```
```suggestion
        // Run the Callable on all members in parallel.
```

angelasimms commented on compute/standard-deviation/client/src/main/java/your/company/name/TotalDifferenceSquaredCallable.java:
```diff
@@ -45,7 +45,7 @@ public TotalDifferenceSquaredCallable(String arg0, double arg1) {
     /**
      * <p>The "{@code hazelcastInstance}" field is set remotely,
-     * on each host where the callable is run.
+     * on each member where the callable is run.
```
```suggestion
     * on each member where the Callable is run.
```

angelasimms commented on compute/standard-deviation/cluster-side/src/main/java/your/company/name/TotalDifferenceSquaredCallable.java:
```diff
@@ -0,0 +1,81 @@
+import com.hazelcast.map.IMap;
+
+/**
+ * <p>A callable to calculate the total squared differences for some
```
```suggestion
 * <p>A Callable to calculate the total squared differences for some
```

angelasimms commented on compute/standard-deviation/cluster-side/src/main/java/your/company/name/TotalDifferenceSquaredCallable.java:
```diff
@@ -0,0 +1,81 @@
+ * We could use {@code stream()} rather than a {@code for-loop} but the
+ * result would be the same.
+ * </p>
+ * <p>The difference though is that this callable runs multiple times,
```
```suggestion
 * <p>The difference though is that this Callable runs multiple times,
```

angelasimms commented on compute/standard-deviation/cluster-side/src/main/java/your/company/name/TotalDifferenceSquaredCallable.java:
```diff
@@ -0,0 +1,81 @@
+
+    /**
+     * <p>The "{@code hazelcastInstance}" field is set remotely,
+     * on each member where the callable is run.
```
```suggestion
     * on each member where the Callable is run.
```

angelasimms commented on compute/standard-deviation/cluster-side/src/main/java/your/company/name/TotalDifferenceSquaredCallable.java:
```diff
@@ -0,0 +1,81 @@
+     * on each member where the callable is run.
+     * </p>
+     *
+     * @param arg0 Injected by Hazelcast when the callable runs.
```
```suggestion
     * @param arg0 Injected by Hazelcast when the Callable runs.
```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,332 @@
+:page-lang: java
+:page-est-time: 30 mins
+:page-serverless: true
+:description: pass:q[In this tutorial, you'll build an application that calculates the standard deviation of customer satifaction scores, using code that's executed in {hazelcast-cloud}.]
```
```suggestion
:description: pass:q[In this tutorial, you'll build an application that calculates the standard deviation of customer satisfaction scores, using code that's executed in {hazelcast-cloud}.]
```

JakeSCahill commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,332 @@
+
+The code is separated into two parts:
+
+- The `client` directory contains the client applications that connect to the cluster and run both the junior developer code and the senior developer code.
```
```suggestion
- The `client` directory contains the client application that connects to the cluster and runs both the junior developer code and the senior developer code.
```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,332 @@
+
+== Context
+
+Knowing the center of a dataset (average, mean, or median) such as customer satisfaction scores doesn't provide enough insight into the best and worse scores. To find out how the customer satisfaction scores are spread out, you can use link:https://en.wikipedia.org/wiki/Standard_deviation[standard deviation].
```
```suggestion
Knowing the center of a dataset (average, mean, or median) doesn't provide enough insight into the best and worse scores. To find out how the customer satisfaction scores are spread out, it is better to use link:https://en.wikipedia.org/wiki/Standard_deviation[standard deviation].
```

JakeSCahill commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,332 @@
+include::standard-deviation:example$client/src/main/java/your/company/name/SeniorDeveloper.java[tag=average]
+----
+
+Hazelcast provides built-in functions for this type of calculation.
```
```suggestion
```

JakeSCahill commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,331 @@
+.Client `SeniorDeveloper.java`
+[%collapsible]
+====
+In this class, the average is calculated in one line:
```
```suggestion
In this class, the average is calculated in one line. Hazelcast provides built-in functions for this type of calculation.
```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,332 @@
+|16
+
+|===
+
```
```suggestion
Examples:
```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,332 @@
+
+The running total for the second set is 16 and there were 5 numbers. 16 divided by 5 is 3.2. The square root of 3.2 gives the standard deviation of *1.78*.
+
+This proves that the first set of numbers doesn't vary as much as the second set.
```
```suggestion
Standard proves that the first set of numbers doesn't vary as much as the second set.
```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,332 @@
+
+In this tutorial, you'll deploy the following to {hazelcast-cloud}:
+
+- A `Customer` class that you'll use to store `Customer` objects in a map. The objects will include a customer satisfaction score.
```
```suggestion
- A `Customer` class to store `Customer` objects in a map. The objects will include a customer satisfaction score.
```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,332 @@
+In this tutorial, you'll deploy the following to {hazelcast-cloud}:
+
+- A `Customer` class that you'll use to store `Customer` objects in a map. The objects will include a customer satisfaction score.
+- An EntryProcessor that the cluster will use to calculate the standard deviation of all customer satisfaction scores in the map.
```
```suggestion
- An entry processor that the cluster will use to calculate the standard deviation of all customer satisfaction scores in the map.
```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,332 @@
+
+You'll then use a client to connect to the cluster, load some `Customer` objects into a map, and calculate the standard deviation of all customer satisfaction scores.
+
+The client code shows both the junior developer and senior developer approach. The junior developer approach does the processing on the client-side. The senior developer approach is more efficient because it offloads the processing to {hazelcast-cloud}, using the EntryProcessor.
```
```suggestion
The client code shows both the junior developer and senior developer approach. The junior developer approach does the processing on the client-side. The senior developer approach is more efficient because it offloads the processing to {hazelcast-cloud} using the entry processor.
```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,332 @@
+- A xref:cloud:ROOT:create-serverless-cluster.adoc[Serverless cluster].
+- link:https://git-scm.com/book/en/v2/Getting-Started-Installing-Git[Git]
+- link:https://maven.apache.org/install.html[Maven]
+- JDK 8, 9, or 11 installed and set up as the `JAVA_HOME` environment variable.
```
Not sure of the wording on this.

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,332 @@
+[[step-one]]
+== Step 1. Clone the Sample Project
+
+All the code for this project is available on GitHub. In this step, you'll clone the project and learn how it works.
```
```suggestion
In this step, you'll clone the sample code from GitHub and learn how it works.
```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,332 @@
+--
+====
+
+The sample code for this tutorial is in the link:{github-directory}[`hazelcast-cloud-code-samples/compute/standard-deviation`] directory.
```
```suggestion
The sample code is in the link:{github-directory}[`hazelcast-cloud-code-samples/compute/standard-deviation`] directory.
```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,332 @@
+The code is separated into two parts:
+
+- The `client` directory contains the client applications that connect to the cluster and run both the junior developer code and the senior developer code.
+- The `cluster-side` directory contains the classes that need to be uploaded to the cluster to allow the cluster to store and process (only for the senior developer code) the customer satisfaction scores.
```
I'm not sure which part of this statement is only for the senior developer

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,332 @@
+- The `client` directory contains the client applications that connect to the cluster and run both the junior developer code and the senior developer code.
+- The `cluster-side` directory contains the classes that need to be uploaded to the cluster to allow the cluster to store and process (only for the senior developer code) the customer satisfaction scores.
+
+The important parts to understand are in the differences between the junior developer code and the senior developer code.
```
```suggestion
The important parts to understand are the differences between the junior developer code and the senior developer code.
```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,332 @@
+
+The junior developer code must request five integers from the cluster in order to run the calculation. The senior developer code runs the calculation on the cluster and returns one double across the network. So the choice is five network transfers or one.
+
+Now imagine the you have 1,000,000 `Customer` objects stored on two Hazelcast members. One member could handle this volume but wouldn't be resilient to failure. So the junior developer approach now requests 1,000,000 integers from the cluster across the network and the senior developer approach retrieves two doubles, one from each member.
```
```suggestion
Now imagine that you have 1,000,000 `Customer` objects stored on two Hazelcast members. One member could handle this volume but wouldn't be resilient to failure. So the junior developer approach now requests 1,000,000 integers from the cluster across the network and the senior developer approach retrieves two doubles, one from each member.
```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,332 @@
+
+This code is easy to understand, which is good for maintenance. However, it has three flaws:
+
+- This code requires every customer record to be moved from where the data is stored (in this case the Hazelcast cluster) to where the calculation is run. Even if a projection were added, this is very heavy on the network.
```
```suggestion
- The code requires every customer record to be moved from where the data is stored, in this case the Hazelcast cluster, to where the calculation is run. Even if a projection were added, this approach places a heavy load on the network.
```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,332 @@
+include::standard-deviation:example$client/src/main/java/your/company/name/SeniorDeveloper.java[tag=stdev]
+----
+
+The client uses the ExecutorService API to run a `Callable` task on each member in the cluster. The `Callable` class is named `TotalDifferenceSquaredCallable.java` and it runs the calculation on the Hazelcast member where it is invoked.
```
```suggestion
The client uses the ExecutorService API to run a Callable task on each member in the cluster. The Callable class is named `TotalDifferenceSquaredCallable.java` and it runs the calculation on the Hazelcast member where it is invoked.
```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,332 @@
+
+The client uses the ExecutorService API to run a `Callable` task on each member in the cluster. The `Callable` class is named `TotalDifferenceSquaredCallable.java` and it runs the calculation on the Hazelcast member where it is invoked.
+
+Each member calculates a sub-total only for the data that member owns. If you have two members, the calculation takes half the time. If you have ten members, the calculation takes a tenth of the time. Each member runs it's calculation independently, the runtime is dependent on how much data each member hosts rather than how much data as a whole exists.
```
```suggestion
Each member calculates a subtotal only for the data that member owns. If you have two members, the calculation takes half the time. If you have ten members, the calculation takes a tenth of the time. Each member runs its calculation independently, the runtime is dependent on how much data each member hosts rather than how much data exists as a whole.
```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,332 @@
+
+Each member in the cluster only needs to work with the keys, and therefore entries, that it owns. As a result, members don't have to do any network calls to retrieve the keys from elsewhere.
+
+Finally, the task is submitted across each member in the cluster, which results in a collection of `Future` objects. One `Future` is returned for each member that runs the task, and the task needs to wait for them all to finish. This is not exactly difficult, and since all members will have roughly the same amount of data, you can assume execution time is equal and iterate across this collection  ...
```
```suggestion
Finally, the task is submitted across each member in the cluster, which results in a collection of `Future` objects. One `Future` is returned for each member that runs the task, and the task needs to wait for them all to finish. This is not exactly difficult, since all members will have roughly the same amount of data. You can assume that the execution time is the same for all members and iterate across this collection running `Future.get()` to get the subtotal from each member.
```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
````diff
@@ -0,0 +1,332 @@
+mvn clean package hazelcast-cloud:deploy
+```
+
+You should see that the file was uploaded and is ready to be used:
````
```suggestion
The file is uploaded and is ready to be used:
```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
````diff
@@ -0,0 +1,332 @@
+mvn clean compile exec:java@standard-deviation
+```
+
+You should see that the client connected to the cluster and calculated the standard deviation, using both the junior developer code and the senior developer code.
````
```suggestion
The client connects to the cluster and calculates the standard deviation using both the junior developer code and the senior developer code.
```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,332 @@
+
+In this tutorial, you learned how to do the following:
+
+- Use an EntryProcessor to process data in parallel on {hazelcast-cloud}.
```
```suggestion
- Use an entry processor to process data in parallel on {hazelcast-cloud}.
```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,332 @@
+In this tutorial, you learned how to do the following:
+
+- Use an EntryProcessor to process data in parallel on {hazelcast-cloud}.
+- Deploy cluster-side modules to a {hazelcast-cloud} cluster, using the {hazelcast-cloud} Maven plugin.
```
```suggestion
- Deploy cluster-side modules to a {hazelcast-cloud} cluster using the {hazelcast-cloud} Maven plugin.
```

JakeSCahill commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,331 @@
+The code is separated into two parts:
+
+- The `client` directory contains the client application that connects to the cluster and runs both the junior developer code and the senior developer code.
+- The `cluster-side` directory contains the classes that need to be uploaded to the cluster to allow the cluster to store and process (only for the senior developer code) the customer satisfaction scores.
```
```suggestion
- The `cluster-side` directory contains the classes that need to be uploaded to the cluster to allow the cluster to store and process the customer satisfaction scores. The customer satisfaction scores are processed by the cluster only in the senior developer code.
```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,331 @@
+|16
+
+|===
+Examples:
```
```suggestion
Examples:

```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,331 @@
+
+|===
+Examples:
+The running total for the first set is 10 and there were 5 numbers. 10 divided by 5 is 2. The square root of 2 gives the standard deviation of *1.41*.
```
```suggestion
- The running total for the first set is 10 and there were 5 numbers. 10 divided by 5 is 2. The square root of 2 gives the standard deviation of *1.41*.
```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,331 @@
+|===
+Examples:
+The running total for the first set is 10 and there were 5 numbers. 10 divided by 5 is 2. The square root of 2 gives the standard deviation of *1.41*.
+
```
```suggestion
```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,331 @@
+Examples:
+The running total for the first set is 10 and there were 5 numbers. 10 divided by 5 is 2. The square root of 2 gives the standard deviation of *1.41*.
+
+The running total for the second set is 16 and there were 5 numbers. 16 divided by 5 is 3.2. The square root of 3.2 gives the standard deviation of *1.78*.
```
```suggestion
- The running total for the second set is 16 and there were 5 numbers. 16 divided by 5 is 3.2. The square root of 3.2 gives the standard deviation of *1.78*.
```

angelasimms reviewed (commented):
Final suggestions.

alaktionov reviewed (commented):
1. 

![image](https://user-images.githubusercontent.com/6086036/184099489-1501371c-2dbf-4d98-9eb0-7cb6e6769e95.png)

2. 

![image](https://user-images.githubusercontent.com/6086036/184099873-ba2d188b-cd2d-44e5-9a56-bd509da989b6.png)

3.

![image](https://user-images.githubusercontent.com/6086036/184101859-7ccbc9d1-6943-4385-a17c-3cc1c3dfec6b.png)

4. 

![image](https://user-images.githubusercontent.com/6086036/184103604-9cd91719-7ba7-401c-a748-85af8c180794.png)

5. 
![image](https://user-images.githubusercontent.com/6086036/184105462-6a494b5c-e8ad-4c57-89f7-0b116645b1c3.png)

JakeSCahill commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,332 @@
+|4
+
+|1
+|1
```
```suggestion
|2
```

JakeSCahill commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -0,0 +1,332 @@
+|Element|Location in {hazelcast-cloud} console
+
+|`<clusterName>`
+|Go to *Connect your Application*. The cluster name is listed in a table at the bottom of the modal window.
```
```suggestion
|Go to *Connect Client*. The cluster name is listed in a table at the bottom of the modal window.
```

JakeSCahill commented:
Thanks @alaktionov 

1. Fixed
2. Will work when we merge
3. Fixed
4. Added this detail to README for Hazelcast devs
5. I just tested and it worked. Can you confirm that you ran the command from the `client` directory?

JakeSCahill commented:
@angelasimms did you test the code in this tutorial? If not, could you try it to see if you get the same error?

angelasimms commented:
> @angelasimms did you test the code in this tutorial? If not, could you try it to see if you get the same error?

@JakeSCahill. No I didn't test the code. I'll see if I can reproduce this.

JakeSCahill commented:
Looks like we were missing a few important steps. The error is thrown because we need to add the keystore files to the `client/src/main/resources` directory. I've added some [extra details](https://github.com/hazelcast/hazelcast-cloud-code-samples/pull/27/commits/5c6cd322374abbd3d94b1d4589abb810213b145d).

JakeSCahill commented on compute/standard-deviation/README.adoc:
````diff
@@ -0,0 +1,58 @@
+. Execute the client application.
++
+```bash
+mvn clean compile exec:java@standard-deviation
````
```suggestion
mvn clean compile exec:java@standard-deviation -Dexec.cleanupDaemonThreads=false
```

JakeSCahill commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
````diff
@@ -0,0 +1,342 @@
+. Execute the client.
++
+```bash
+mvn clean compile exec:java@standard-deviation
````
```suggestion
mvn clean compile exec:java@standard-deviation -Dexec.cleanupDaemonThreads=false
```

angelasimms commented on compute/standard-deviation/docs/modules/ROOT/pages/compute-standard-deviation.adoc:
```diff
@@ -269,6 +269,14 @@ The file is uploaded and is ready to be used:
+
+. Go to *Connect Client* > *Advanced setup*, and click *Download keystore file*.
+
+. Extract the files and copy the `client.keystore` and `client.truststore` files to the `client/src/main/resources` directory.
```
Minor point. I had to create the resources directory. It didn't previously exist.

JakeSCahill replied:
Thanks - I added a file to the directory so that we can commit it.

angelasimms reviewed (commented):
Hi Jake, Just a suggestion for a minor update to the next step instructions.
